File: service/javadsl/api/src/main/java/com/lightbend/lagom/javadsl/api/broker/kafka/PartitionKeyStrategy.java
/*
 * Copyright (C) Lightbend Inc. <https://www.lightbend.com>
 */

package com.lightbend.lagom.javadsl.api.broker.kafka;

/** Defines an algorithm for producing a key from a Message payload. */
@FunctionalInterface
public interface PartitionKeyStrategy<Payload> {

  /**
   * Computes a key from a message payload. The key is used to decide on what topic's partition a
   * message should be published to.
   *
   * @param payload The message payload to publish into a Kafka topic.
   * @return A key used to decide on what topic's partition the passed message is published to.
   */
  String computePartitionKey(Payload payload);
}

File: service/scaladsl/api/src/main/scala/com/lightbend/lagom/scaladsl/api/broker/kafka/PartitionKeyStrategy.scala
/*
 * Copyright (C) Lightbend Inc. <https://www.lightbend.com>
 */

package com.lightbend.lagom.scaladsl.api.broker.kafka

/**
 * Defines an algorithm for producing a key from a Message.
 */
trait PartitionKeyStrategy[-Payload] {

  /**
   * Computes a key from a message payload.
   * The key is used to decide on what topic's partition a message should be published to.
   *
   * @param message The message to publish into a Kafka topic.
   * @return A key used to decide on what topic's partition the passed message is published to.
   */
  def computePartitionKey(message: Payload): String
}

object PartitionKeyStrategy {

  /**
   * Create a partition key strategy.
   */
  def apply[Payload](f: Payload => String): PartitionKeyStrategy[Payload] = new PartitionKeyStrategy[Payload] {
    override def computePartitionKey(message: Payload): String = f(message)
  }
}
